README.md
DayZ Linux CLI Launcher
====

## About

This is an experimental launcher script for DayZ standalone on Linux when running the game via Proton.

Proton is currently unable to start the game's own regular launcher application which sets up mods and launch parameters for the game client. The game however does work fine when launching the client directly, so mods can be set up and configured manually, which is what this script does, similar to what the launcher would do.

Automatic Steam workshop mod downloads are currently unsupported due to a limitation of Steam's CLI. Workshop mods will therefore need to be subscribed manually via the web browser. A URL for each missing mod will be printed to the output.

Please see the "Install DayZ" section down below on how to get the game running on Linux.

## Usage

```
Usage: dayz-launcher.sh [OPTION]... [MODID]... [-- [GAME-PARAM]...]

Automatically set up mods for DayZ, launch the game and connect to a server,
or print the game's -mod command line argument for custom configuration.

Command line options:

  -h
  --help
    Print this help text.

  -d
  --debug
    Print debug messages to output.

  --steam <"" | flatpak | /path/to/steam/executable>
    If set to flatpak, use the flatpak version of Steam (com.valvesoftware.Steam).
    Steam needs to already be running in the flatpak container.
    Default is: "" (automatic detection - prefers flatpak if available)

  -l
  --launch
    Launch DayZ after resolving and setting up mods instead of
    printing the game's -mod command line argument.
    Any custom game parameters that come after the first double-dash (--) will
    be appended to the overall launch command line. This implies --launch.

  -n <name>
  --name <name>
    Set the profile name when launching the game via --launch.
    Some community servers require a profile name when trying to connect.

  -s <address[:port]>
  --server <address[:port]>
    Retrieve a server's mod list and add it to the remaining input.
    Uses the dayzsalauncher.com JSON API.
    If --launch is set, it will automatically connect to the server.
    The optional port is the server's game port. Default is: 2302

  -p <port>
  --port <port>
    The server's query port, not to be confused with the server's game port.
    Default is: 27016

Environment variables:

  STEAM_ROOT
    Set a custom path to Steam's root directory. Default is:
    ${XDG_DATA_HOME:-${HOME}/.local/share}/Steam
    which defaults to ~/.local/share/Steam

    If the flatpak package is being used, then the default is:
    ~/.var/app/com.valvesoftware.Steam/data/Steam

    If the game is stored in a different Steam library directory, then this
    environment variable needs to be set/changed.

    For example, if the game has been installed in the game library located in
      /media/games/SteamLibrary/steamapps/common/DayZ
    then the STEAM_ROOT env var needs to be set like this:
      STEAM_ROOT=/media/games/SteamLibrary
```

## Examples

```sh
# configure mods by ID
./dayz-launcher.sh -d 123 456 789

# retrieve mods list from server, configure mods and launch the game
./dayz-launcher.sh -d -l -s address:gameport -p queryport -n nickname

# configure mods by ID and connect
./dayz-launcher.sh -d 123 456 789 -- -connect=address -name=nickname

# run game from a different Steam games library path
STEAM_ROOT=/media/games/SteamLibrary ./dayz-launcher.sh -l -s address
```

## Known issues

### Third party server query API

Server data is queried via the third-party dayzsalauncher.com JSON API when using the `--server` parameter. Previously, this was done via the daemonforge.dev JSON API, which unfortunately wasn't perfectly reliable. Please try running the launcher again if the query times out, and if the data returned by the server query API doesn't reflect the server's actual mod IDs, then custom mod IDs will need to be set as launch arguments.

### Steam doesn't launch the game

Sometimes the game doesn't launch despite the launcher correctly passing the right parameters to Steam. This is usually caused by a broken singleton process detection of Steam, where it passes the game's launch parameters to the original Steam process and then terminates. Restarting Steam and re-running the launcher fixes the issue.

One of the main reasons why this can happen is an [unresolved bug in the Steam client](https://github.com/ValveSoftware/steam-for-linux/issues/5753) which causes Steam to become unresponsive when launching it via `-applaunch` and passing arguments that reach a certain threshold in length. This problem occurs when trying to join servers which require loading lots of mods and which therefore increase the length of the launch parameters. In order to overcome this limitation, mod directories have been shrinked as much as possible in the `0.5.0` release, so that the launch parameters can be kept as short as possible. This comes at the cost of not being able to tell mod directories apart in the DayZ game directory, at least as a human, but this should solve the issue for the vast majority of servers. Please report any further issues on the issue tracker. Thank you.

### Case sensitivity of mod-internal file paths

Due to mods being developed on Windows and Windows not differentiating between file paths with different cases, mod developers sometimes don't care about case sensitivity of file names and directories. This can lead to loading issues on Linux when a file system without case insensitivity support is being used.

In case of ext4, which is the default file system on most distros, case insensitivity has only been implemented in kernel 5.2, which was released in July 2019 (later improved for encrypted file systems in kernel 5.13, released in June 2021). The requirements for enabling case insensitivity are an ext4 filesystem created with at least kernel version 5.2 while the `casefold` option was set, as well as the `+F` flag being set on one of the parent directories.

Please refer to your distro's documentation or search engine of choice for how to properly enable case folding/insensitivity on your file system and whether it's supported.

## Future ideas

- Rewrite in Python
- Don't use a third party server query API and query the server directly
- Install mods automatically  
  Unfortunately, Steam doesn't support downloading workshop mods via the CLI and only the `steamcmd` CLI utility seems to be able to do this from a command line shell context, but this requires a Steam login via CLI parameters, which is a bit unpractical.
- If possible, resolve mod dependencies

## Install

To install the launcher script, simply clone the git repository:

```sh
git clone https://github.com/bastimeyer/dayz-linux-cli-launcher.git
cd dayz-linux-cli-launcher
./dayz-launcher.sh ...
```

or download the raw script file from GitHub and make it executable (check the script file contents first before running it):

```sh
curl -SL -o dayz-launcher.sh 'https://github.com/bastimeyer/dayz-linux-cli-launcher/raw/master/dayz-launcher.sh'
chmod +x dayz-launcher.sh
./dayz-launcher.sh ...
```

## Install DayZ

[Support for BattlEye anti-cheat for Proton on Linux has been officially announced by Valve on 2021-12-03.][battleye-announcement]

In order to get the game running on Linux, you first have to install the Steam beta client (see Steam's settings menu). Then install `Proton Experimental` and the `Proton BattlEye Runtime` (filter by "tools" in your games library). After that, set the "Steam play compatibility tool" for DayZ to "Proton Experimental" (right-click the game and go to properties).

### Important notes

In order for the game to actually run on Linux via Proton, the [`vm.max_map_count`][vm.max_map_count] kernel parameter needs to be increased, because otherwise the game will freeze while loading the main menu or after playing for a couple of minutes. Some custom kernels like TK-Glitch for example already increase this value from its [default value of `64*1024-6`][vm.max_map_count-default] to [`512*1024`][tkg-kernel-patch], but even this won't work reliably. Increasing it to `1024*1024` seems to work.

```sh
​sudo sysctl -w vm.max_map_count=1048576
```

Or apply it permanently:

```sh
​echo 'vm.max_map_count=1048576' | sudo tee /etc/sysctl.d/vm.max_map_count.conf
```


  [battleye-announcement]: https://store.steampowered.com/news/group/4145017/view/3104663180636096966
  [vm.max_map_count]: https://github.com/torvalds/linux/blob/v5.15/Documentation/admin-guide/sysctl/vm.rst#max_map_count
  [vm.max_map_count-default]: https://github.com/torvalds/linux/blob/v5.15/include/linux/mm.h#L185-L202
  [tkg-kernel-patch]: https://github.com/Frogging-Family/linux-tkg/blob/db405096bd7fb52656fc53f7c5ee87e7fe2f99c9/linux-tkg-patches/5.15/0003-glitched-base.patch#L477-L534


CHANGELOG.md
Changelog
====

## 0.5.3 - 2022-06-18

- Fixed mods failing to load when the base64 modID string included a `+` character ([#17](https://github.com/bastimeyer/dayz-linux-cli-launcher/pull/17))

## 0.5.2 - 2022-05-26

- Fixed default game port not being added to the server address if the port is missing

## 0.5.1 - 2022-05-25

- Fixed error when querying unmodded servers

## 0.5.0 - 2022-04-10

- Shortened the mod directory link names in the DayZ game directory.  
  This fixes the issue of Steam not launching the game and becoming unresponsive to further launch attempts.  
  See [#9](https://github.com/bastimeyer/dayz-linux-cli-launcher/issues/9) and [ValveSoftware/steam-for-linux#5753](https://github.com/ValveSoftware/steam-for-linux/issues/5753).

## 0.4.0 - 2022-02-19

- Switched from daemonforge.dev JSON API to dayzsalauncher.com JSON API

## 0.3.0 - 2021-12-21

- Added support for optional custom game launch parameters
- Added examples and known issues to readme

## 0.2.2 - 2021-12-17

- Fixed mod symlink path check

## 0.2.1 - 2021-12-17

- Added `STEAM_ROOT` example to `--help`.
- Added output messages when not doing anything (missing mod-IDs, server address or launch parameter).

## 0.2.0 - 2021-12-15

- Changed mod symlinks to the `modid-modname` format

## 0.1.2 - 2021-12-08

- Fixed launcher not working when `--debug` was not set due to incorrect return value of debug log function

## 0.1.1 - 2021-12-08

- Fixed `stderr` redirection of various commands

## 0.1.0 - 2021-12-06

- Implemented flatpak support. Tries to use flatpak by default, if available.
- Added `--steam` parameter for overriding flatpak mode or choosing a different steam executable.

## 0.0.1 - 2021-12-06

- Initial release


dayz-mods.sh
#!/usr/bin/env bash
set -eo pipefail

SELF=$(basename "$(readlink -f "${0}")")

[[ -z "${STEAM_ROOT}" ]] && STEAM_ROOT="${XDG_DATA_HOME:-${HOME}/.local/share}/Steam"
STEAM_ROOT="${STEAM_ROOT}/steamapps"

DAYZ_ID=221100
DIR_WORKSHOP="${STEAM_ROOT}/workshop/content/${DAYZ_ID}"
DIR_DAYZ="${STEAM_ROOT}/common/DayZ"

API_URL="https://api.daemonforge.dev/server/@ADDRESS@/@PORT@/full"
API_PARAMS=(
  -sSL
  -m 10
  -H "Referer: https://daemonforge.dev/"
  -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.45 Safari/537.36"
)

WORKSHOP_URL="https://steamcommunity.com/sharedfiles/filedetails/?id=@ID@"

DEBUG=0
LAUNCH=0
SERVER=""
PORT="27016"
NAME=""
INPUT=()
MODS=()

declare -A DEPS=(
  [gawk]=gawk
  [curl]=curl
  [jq]=jq
  [steam]=steam
)


print_help() {
  cat <<EOF
Usage: ${SELF} [OPTION]... [MODID]...

Automatically set up mods for the DayZ client
and print the game's -mod command line argument.

Command line options:

  -h
  --help
    Print this help text.

  -d
  --debug
    Print debug messages to output.

  -l
  --launch
    Launch DayZ after resolving and setting up mods
    instead of printing the game's -mod command line argument.

  -n <name>
  --name <name>
    Set the profile name when launching the game via --launch.
    Some community servers require a profile name when trying to connect.

  -s <address[:port]>
  --server <address[:port]>
    Retrieve a server's mod list and add it to the remaining input.
    Uses the daemonforge.dev DayZ server JSON API.
    If --launch is set, it will automatically connect to the server.

  -p <port>
  --port <port>
    The server's query port (not to be confused with the server's game port).
    Default is: 27016

Environment variables:

  STEAM_ROOT
    Set a custom path to Steam's root directory. Default is:
    \${XDG_DATA_HOME:-\${HOME}/.local/share}/Steam
    which defaults to ~/.local/share/Steam

    If the game is stored in a different Steam library directory, then this
    environment variable needs to be set/changed.
EOF
}


while (( "$#" )); do
  case "${1}" in
    -h|--help)
      print_help
      exit
      ;;
    -d|--debug)
      DEBUG=1
      ;;
    -l|--launch)
      LAUNCH=1
      ;;
    -s|--server)
      SERVER="${2}"
      shift
      ;;
    -p|--port)
      PORT="${2}"
      shift
      ;;
    -n|--name)
      NAME="${2}"
      shift
      ;;
    *)
      INPUT+=("${1}")
      ;;
  esac
  shift
done


# ----


err() {
  echo >&2 "[${SELF}][error] ${@}"
  exit 1
}

msg() {
  echo "[${SELF}][info] ${@}"
}

debug() {
  [[ ${DEBUG} == 1 ]] && echo "[${SELF}][debug] ${@}"
}

check_dir() {
  [[ -d "${1}" ]] || err "Invalid/missing directory: ${1}"
}


# ----


check_deps() {
  for dep in "${!DEPS[@]}"; do
    command -v "${dep}" 2>&1 >/dev/null || err "${DEPS["${dep}"]} is missing. Aborting."
  done
}

query_server_api() {
  [[ -z "${SERVER}" ]] && return

  local query
  local response
  msg "Querying API for server: ${SERVER%:*}:${PORT}"
  query="$(sed -e "s/@ADDRESS@/${SERVER%:*}/" -e "s/@PORT@/${PORT}/" <<< "${API_URL}")"
  debug "Querying ${query}"
  response="$(curl "${API_PARAMS[@]}" "${query}")"
  debug "Parsing API response"
  jq -e ".mods[]" 2>&1 >/dev/null <<< "${response}" || err "Missing mods data from API response"

  INPUT+=( $(jq -r ".mods[] | select(.app_id == ${DAYZ_ID}) | .id" <<< "${response}") )
}

setup_mods() {
  local missing=0
  for modid in "${INPUT[@]}"; do
    local modpath="${DIR_WORKSHOP}/${modid}"
    if ! [[ -d "${modpath}" ]]; then
      missing=1
      msg "Missing mod directory for: ${modid}"
      msg "Subscribe the mod here: $(sed -e "s/@ID@/${modid}/" <<< "${WORKSHOP_URL}")"
      continue
    fi

    local modmeta="${modpath}/meta.cpp"
    [[ -f "${modmeta}" ]] || err "Missing mod metadata for: ${modid}"

    local modname="$(gawk 'match($0,/name\s*=\s*"(.+)"/,m){print m[1];exit}' "${modmeta}")"
    [[ -n "${modname}" ]] || err "Missing mod name for: ${modid}"
    debug "Mod ${modid} found: ${modname}"
    modname="${modname//\'/}"

    if ! [[ -L "${DIR_DAYZ}/@${modname}" ]]; then
      msg "Creating mod symlink for: ${modname}"
      ln -sr "${modpath}" "${DIR_DAYZ}/@${modname}"
    fi

    MODS+=("@${modname}")
  done

  return ${missing}
}


main() {
  check_deps
  check_dir "${DIR_DAYZ}"
  check_dir "${DIR_WORKSHOP}"

  query_server_api
  setup_mods || exit 1

  local mods="$(IFS=";"; echo "${MODS[*]}")"

  if [[ "${LAUNCH}" == 1 ]]; then
    local cmdline=()
    [[ -n "${mods}" ]] && cmdline+=("-mod=${mods}")
    [[ -n "${SERVER}" ]] && cmdline+=("-connect=${SERVER}" -nolauncher -world=empty)
    [[ -n "${NAME}" ]] && cmdline+=("-name=${NAME}")
    ( set -x; steam -applaunch "${DAYZ_ID}" "${cmdline[@]}"; )
  elif [[ -n "${mods}" ]]; then
    msg "Add this to your game's launch options, including the quotes:"
    echo "\"-mod=${mods}\""
  fi
}

main


dayz-launcher.sh
#!/usr/bin/env bash
set -eo pipefail

SELF=$(basename "$(readlink -f "${0}")")

DAYZ_ID=221100

DEFAULT_GAMEPORT=2302
DEFAULT_QUERYPORT=27016

FLATPAK_STEAM="com.valvesoftware.Steam"
FLATPAK_PARAMS=(
  --branch=stable
  --arch=x86_64
  --command=/app/bin/steam-wrapper
)

API_URL="https://dayzsalauncher.com/api/v1/query/@ADDRESS@/@PORT@"
API_PARAMS=(
  -sSL
  -m 10
  -H "User-Agent: dayz-linux-cli-launcher"
)

WORKSHOP_URL="https://steamcommunity.com/sharedfiles/filedetails/?id=@ID@"


DEBUG=0
LAUNCH=0
STEAM=""
SERVER=""
PORT="${DEFAULT_QUERYPORT}"
NAME=""
INPUT=()
MODS=()
PARAMS=()

declare -A DEPS=(
  [gawk]="required for parsing the mod metadata"
  [curl]="required for querying the server API"
  [jq]="required for parsing the server API's JSON response"
)


print_help() {
  cat <<EOF
Usage: ${SELF} [OPTION]... [MODID]... [-- [GAME-PARAM]...]

Automatically set up mods for DayZ, launch the game and connect to a server,
or print the game's -mod command line argument for custom configuration.

Command line options:

  -h
  --help
    Print this help text.

  -d
  --debug
    Print debug messages to output.

  --steam <"" | flatpak | /path/to/steam/executable>
    If set to flatpak, use the flatpak version of Steam (${FLATPAK_STEAM}).
    Steam needs to already be running in the flatpak container.
    Default is: "" (automatic detection - prefers flatpak if available)

  -l
  --launch
    Launch DayZ after resolving and setting up mods instead of
    printing the game's -mod command line argument.
    Any custom game parameters that come after the first double-dash (--) will
    be appended to the overall launch command line. This implies --launch.

  -n <name>
  --name <name>
    Set the profile name when launching the game via --launch.
    Some community servers require a profile name when trying to connect.

  -s <address[:port]>
  --server <address[:port]>
    Retrieve a server's mod list and add it to the remaining input.
    Uses the dayzsalauncher.com DayZ server JSON API.
    If --launch is set, it will automatically connect to the server.
    The optional port is the server's game port. Default is: ${DEFAULT_GAMEPORT}

  -p <port>
  --port <port>
    The server's query port, not to be confused with the server's game port.
    Default is: ${DEFAULT_QUERYPORT}

Environment variables:

  STEAM_ROOT
    Set a custom path to Steam's root directory. Default is:
    \${XDG_DATA_HOME:-\${HOME}/.local/share}/Steam
    which defaults to ~/.local/share/Steam

    If the flatpak package is being used, then the default is:
    ~/.var/app/${FLATPAK_STEAM}/data/Steam

    If the game is stored in a different Steam library directory, then this
    environment variable needs to be set/changed.

    For example, if the game has been installed in the game library located in
      /media/games/SteamLibrary/steamapps/common/DayZ
    then the STEAM_ROOT env var needs to be set like this:
      STEAM_ROOT=/media/games/SteamLibrary
EOF
}


while (( "$#" )); do
  case "${1}" in
    -h|--help)
      print_help
      exit
      ;;
    -d|--debug)
      DEBUG=1
      ;;
    --steam)
      STEAM="${2}"
      shift
      ;;
    -l|--launch)
      LAUNCH=1
      ;;
    -s|--server)
      SERVER="${2}"
      [[ "${SERVER}" = *:* ]] || SERVER="${SERVER}:${DEFAULT_GAMEPORT}"
      shift
      ;;
    -p|--port)
      PORT="${2}"
      shift
      ;;
    -n|--name)
      NAME="${2}"
      shift
      ;;
    --)
      shift
      PARAMS+=("${@}")
      LAUNCH=1
      break
      ;;
    *)
      INPUT+=("${1}")
      ;;
  esac
  shift
done


# ----


err() {
  echo >&2 "[${SELF}][error] ${@}"
  exit 1
}

msg() {
  echo "[${SELF}][info] ${@}"
}

debug() {
  if [[ ${DEBUG} == 1 ]]; then
    echo "[${SELF}][debug] ${@}"
  fi
}

check_dir() {
  debug "Checking directory: ${1}"
  [[ -d "${1}" ]] || err "Invalid/missing directory: ${1}"
}

check_dep() {
  command -v "${1}" >/dev/null 2>&1
}

check_deps() {
  for dep in "${!DEPS[@]}"; do
    check_dep "${dep}" || err "'${dep}' is missing (${DEPS["${dep}"]}). Aborting."
  done
}

check_flatpak() {
  check_dep flatpak \
    && flatpak info "${FLATPAK_STEAM}" >/dev/null 2>&1 \
    && { flatpak ps | grep "${FLATPAK_STEAM}"; } >/dev/null 2>&1
}

dec2base64() {
  echo "$1" \
    | LC_ALL=C gawk '
      {
        do {
          printf "%c", and($1, 255)
          $1 = rshift($1, 8)
        } while ($1 > 0)
      }
    ' \
    | base64 \
    | sed 's|/|-|g; s|+|_|g; s|=||g'
}


# ----


resolve_steam() {
  if [[ "${STEAM}" == flatpak ]]; then
    check_flatpak || err "Could not find a running instance of the '${FLATPAK_STEAM}' flatpak package"
  elif [[ -n "${STEAM}" ]]; then
    check_dep "${STEAM}" || err "Could not find the '${STEAM}' executable"
  else
    msg "Resolving steam"
    if check_flatpak; then
      STEAM=flatpak
    elif check_dep steam; then
      STEAM=steam
    else
      err "Could not find a running instance of the '${FLATPAK_STEAM}' flatpak package or the 'steam' executable"
    fi
  fi
}

query_server_api() {
  [[ -z "${SERVER}" ]] && return

  local query
  local response
  msg "Querying API for server: ${SERVER%:*}:${PORT}"
  query="$(sed -e "s/@ADDRESS@/${SERVER%:*}/" -e "s/@PORT@/${PORT}/" <<< "${API_URL}")"
  debug "Querying ${query}"
  response="$(curl "${API_PARAMS[@]}" "${query}")"
  debug "Parsing API response"
  jq -e '.result.mods | select(type == "array")' >/dev/null 2>&1 <<< "${response}" || err "Missing mods data from API response"
  jq -e '.result.mods[]' >/dev/null 2>&1 <<< "${response}" || { msg "This server is unmodded"; return; }

  INPUT+=( $(jq -r ".result.mods[] | .steamWorkshopId" <<< "${response}") )
}

setup_mods() {
  local dir_dayz="${1}"
  local dir_workshop="${2}"
  local missing=0

  for modid in "${INPUT[@]}"; do
    local modpath="${dir_workshop}/${modid}"
    if ! [[ -d "${modpath}" ]]; then
      missing=1
      msg "Missing mod directory for: ${modid}"
      msg "Subscribe the mod here: $(sed -e "s/@ID@/${modid}/" <<< "${WORKSHOP_URL}")"
      continue
    fi

    local modmeta="${modpath}/meta.cpp"
    [[ -f "${modmeta}" ]] || err "Missing mod metadata for: ${modid}"

    local modname="$(gawk 'match($0,/name\s*=\s*"(.+)"/,m){print m[1];exit}' "${modmeta}")"
    [[ -n "${modname}" ]] || err "Missing mod name for: ${modid}"
    debug "Mod ${modid} found: ${modname}"
    local modlink="@$(dec2base64 "${modid}")"

    if ! [[ -L "${dir_dayz}/${modlink}" ]]; then
      msg "Creating mod symlink for: ${modname} (${modlink})"
      ln -sr "${modpath}" "${dir_dayz}/${modlink}"
    fi

    MODS+=("${modlink}")
  done

  return ${missing}
}

run_steam() {
  if [[ "${STEAM}" == flatpak ]]; then
    ( set -x; flatpak run "${FLATPAK_PARAMS[@]}" "${FLATPAK_STEAM}" "${@}"; )
  else
    ( set -x; steam "${@}"; )
  fi
}


main() {
  check_deps
  resolve_steam

  if [[ "${STEAM}" == flatpak ]]; then
    msg "Using flatpak mode"
  else
    msg "Using non-flatpak mode: ${STEAM}"
  fi

  if [[ -z "${STEAM_ROOT}" ]]; then
    if [[ "${STEAM}" == flatpak ]]; then
      STEAM_ROOT="${HOME}/.var/app/${FLATPAK_STEAM}/data/Steam"
    else
      STEAM_ROOT="${XDG_DATA_HOME:-${HOME}/.local/share}/Steam"
    fi
  fi
  STEAM_ROOT="${STEAM_ROOT}/steamapps"

  local dir_dayz="${STEAM_ROOT}/common/DayZ"
  local dir_workshop="${STEAM_ROOT}/workshop/content/${DAYZ_ID}"
  check_dir "${dir_dayz}"
  check_dir "${dir_workshop}"

  query_server_api
  setup_mods "${dir_dayz}" "${dir_workshop}" || exit 1

  local mods="$(IFS=";"; echo "${MODS[*]}")"

  if [[ "${LAUNCH}" == 1 ]]; then
    local cmdline=()
    [[ -n "${mods}" ]] && cmdline+=("-mod=${mods}")
    [[ -n "${SERVER}" ]] && cmdline+=("-connect=${SERVER}" -nolauncher -world=empty)
    [[ -n "${NAME}" ]] && cmdline+=("-name=${NAME}")
    msg "Launching DayZ"
    run_steam -applaunch "${DAYZ_ID}" "${cmdline[@]}" "${PARAMS[@]}"
  elif [[ -n "${mods}" ]]; then
    msg "Add this to your game's launch options, including the quotes:"
    echo "\"-mod=${mods}\""
  else
    msg "Nothing to do..."
    msg "No mod-ID list, --server address, or --launch parameter set."
    msg "See --help for all the available options."
  fi
}

main
